File: src/main/java/learners/Decider.java
package learners;

import main.InputRow;

import java.io.*;
import java.util.List;

public interface Decider extends Serializable {

	/** Returns the decision for the input value */
	LanguageDecision decide(InputRow row);

	/** Human readable output for debugging. */
	String representation(int numSpaces);

	/** Tests this decider on the input data, and returns the number incorrect / total. */
	default double errorRateUnWeighted(List<InputRow> testingData) {
		double total = testingData.size();
		double correct = testingData.stream()
			.filter(i -> this.decide(i).mostConfidentLanguage().equals(i.outputValue))
			.count();

		return (total - correct) / total;
	}

	/** Writes this object using java serialization to the filename. */
	default void saveTo(String fileName) throws IOException {
		FileOutputStream file = new FileOutputStream(fileName);
		ObjectOutputStream out = new ObjectOutputStream(file);
		out.writeObject(this);
		out.flush();
		out.close();
		file.close();
	}

	/** Loads a decider from the file using java deserialization */
	static Decider loadFromFile(String fileName) throws IOException, ClassNotFoundException {
		FileInputStream file = new FileInputStream(fileName);
		ObjectInputStream in = new ObjectInputStream(file);
		Decider result = (Decider) in.readObject();
		in.close();
		file.close();
		return result;
	}

	/** Loads a decider from the resources file name. */
	static Decider loadFromResources(String resourcesFileName) throws IOException, ClassNotFoundException {
		InputStream file = Decider.class.getClassLoader().getResourceAsStream(resourcesFileName);
		ObjectInputStream in = new ObjectInputStream(file);
		Decider result = (Decider) in.readObject();
		in.close();
		file.close();
		return result;
	}
}

File: src/main/kotlin/com/zachjones/languageclassifier/core/Decider.kt
package com.zachjones.languageclassifier.core

import com.zachjones.languageclassifier.entities.InputRow
import com.zachjones.languageclassifier.entities.LanguageDecision
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable

interface Decider : Serializable {
    /** Returns the decision for the input value  */
    fun decide(row: InputRow): LanguageDecision

    /** Human readable output for debugging.  */
    fun representation(numSpaces: Int): String?

    /** Tests this decider on the input data, and returns the number incorrect / total.  */
    fun errorRateUnWeighted(testingData: List<InputRow>): Double {
        val total = testingData.size.toDouble()
        val correct = testingData.stream()
            .filter { i: InputRow -> decide(i).mostConfidentLanguage() == i.language }
            .count().toDouble()

        return (total - correct) / total
    }

    /** Writes this object using java serialization to the filename.  */
    @Throws(IOException::class)
    fun saveTo(fileName: String) {
        val file = FileOutputStream(fileName)
        val out = ObjectOutputStream(file)
        out.writeObject(this)
        out.flush()
        out.close()
        file.close()
    }

    companion object {
        /** Loads a decider from the file using java deserialization  */
        @Throws(IOException::class, ClassNotFoundException::class)
        fun loadFromFile(fileName: String): Decider {
            val file = FileInputStream(fileName)
            val `in` = ObjectInputStream(file)
            val result = `in`.readObject() as Decider
            `in`.close()
            file.close()
            return result
        }
    }
}
